funsor/delta.py

```diff
@@ -188,7 +67,142 @@ def solve_unary(op, arg, y):
+            return self
+
+        new_terms = sorted(self.terms, key=lambda t: names.index(t[0]))
+        return MultiDelta(new_terms)
+
+    def eager_subs(self, subs):
+        terms = OrderedDict(self.terms)
+        new_terms = terms.copy()
+        log_density = to_funsor(0., self.output)
+        for name, value in subs:
+            if isinstance(value, Variable):
+                new_terms[value.name] = new_terms.pop(name)
+                continue
+
+            if not any(d.dtype == 'real' for side in (value, terms[name][0])
+                       for d in side.inputs.values()):
+                point, point_log_density = new_terms.pop(name)
+                log_density += (value == point).all().log() + point_log_density
+                continue
+
+            # Try to invert the substitution.
+            soln = solve(value, terms[name][0])
+            if soln is None:
+                return None  # lazily substitute
+            new_name, new_point, point_log_density = soln
+            old_point, old_point_density = new_terms.pop(name)
+            new_terms[new_name] = (new_point, old_point_density + point_log_density)
+
+        return MultiDelta(tuple(new_terms.items())) + log_density if new_terms else log_density
+
+    def eager_reduce(self, op, reduced_vars):
+        if op is ops.logaddexp:
+            if reduced_vars - self.fresh and self.fresh - reduced_vars:
+                result = self.eager_reduce(op, reduced_vars & self.fresh) if reduced_vars & self.fresh else self
+                if result is not self:
+                    result = result.eager_reduce(op, reduced_vars - self.fresh) if reduced_vars - self.fresh else self
+                    return result if result is not self else None
+                return None
+
+            result_terms = [(name, (point, log_density)) for name, (point, log_density) in self.terms
+                            if name not in reduced_vars]
+
+            result_terms, scale = [], Number(0)
+            for name, (point, log_density) in self.terms:
+                if name in reduced_vars:
+                    # XXX obscenely wasteful - need a lazy Zero term
+                    if point.inputs:
+                        scale += (point == point).all().log()
+                    if log_density.inputs:
+                        scale += log_density * 0.
+                else:
+                    result_terms.append((name, (point, log_density)))
+
+            result = MultiDelta(tuple(result_terms)) + scale if result_terms else scale
+            return result.reduce(op, reduced_vars - self.fresh)
+
+        if op is ops.add:
+            raise NotImplementedError("TODO Implement ops.add to simulate .to_event().")
+
+        return None  # defer to default implementation
+
+    def unscaled_sample(self, sampled_vars, sample_inputs):
+        # raise NotImplementedError("TODO implement sample for particle indices")
```

Comment:
Maybe we should err if the wrong variables are being sampled?
```py
if not sampled_vars.issubset(self.fresh):
    raise ValueError
```

Reply:
This is identical to the current behavior of `Joint.unscaled_sample`, and I believe it's necessary for `monte_carlo` in its current incarnation to work since it repeatedly `.sample`s until it stops making progress.  We should definitely revisit in Monte Carlo cleanup/refactoring PRs.

Reply:
Well either way, please remove commented code.